### PR TITLE
BE server, Router, Controller

### PR DESCRIPTION
**Pull request by davidzhangnyc**
WHAT DOES THIS PR DO? 
1. Configured errorRouter and errorController to send back a JSON object of the event logs

HOW TO TEST THIS PR:
1. Run Docker Desktop and in the terminal run `minikube start`
2. Create a mongodb instance and create a env file with the URI OR paste it in the URI field in the errorModel.js file (remember to clear the field after testing)
3. Run `npm install` and `nodemon server/server.js`
4. Go to localhost:3000/error to see all the event logs

ADDITIONAL NOTES:
This is what an event log should look like within the array:
{
  "_id": "60e46ee7888d2a1a14e10e6e",
  "namespace": "default",
  "lastSeen": "30m",
  "type": "Normal",
  "reason": "Created",
  "object": "pod/alertmanager-prometheus-kube-prometheus-alertmanager-0",
  "message": "Created container config-reloader",
  "createdAt": "2021-07-06T14:55:35.126Z",
  "updatedAt": "2021-07-06T14:55:35.126Z",
  "__v": 0
}

### CODE REVIEW

**Review by JefZheng — APPROVED** [review]
Recommend to install a JSON formatting extension on your browser to view the information correctly @dlande000 @Hydroelectric29

File: server/models/errorsModel.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  },
+  message: {
+    required: true,
+    type: String
+  },
+}, {
+  timestamps: true,
+})
+
+const Error = mongoose.model('error', errorSchema);
```
Comment (dlande000) [inline server/models/errorsModel.js]: We should rename this to `K8sError` or some similar name—right now, we are overwriting the default JS Error class.

File: server/models/errorsModel.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+module.exports = Error;
```
Comment (dlande000) [inline server/models/errorsModel.js]: Add new line at end of file: https://stackoverflow.com/questions/44704968/visual-studio-code-insert-new-line-at-the-end-of-files

File: server/server.js [hunk 3]
```diff
@@ -10,6 +11,8 @@ app.use(express.urlencoded({ extended: true }));
 app.use(express.static(path.resolve(__dirname + '../src')));
 app.use('/build', express.static(path.join(__dirname + '../build')));
 
+app.use('/error', errorRouter);
```
Comment (dlande000) [inline server/server.js]: Change `error` to `errors` and `errorRouter` to `errorsRouter`: https://blog.dreamfactory.com/best-practices-for-naming-rest-api-endpoints/#:~:text=Use%20nouns%20for%20naming%20URIs,%3A%2F%2Fapi.example.com.&text=In%20general%2C%20using%20plural%20nouns,admin%20for%20the%20administrative%20user).

File: server/server.js [hunk 4]
```diff
@@ -1,5 +1,6 @@
 const express = require('express');
 const path = require('path');
+const errorRouter = require('./routers/errorRouter');
```
Comment (dlande000) [inline server/server.js]: See : https://github.com/oslabs-beta/Kubiquity/pull/12/files#r664688852

File: server/routers/errorRouter.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  // errorController.clearErrors,
+  errorController.queryErrors,
+  errorController.formatErrors,
+  errorController.saveErrors,
+  errorController.getErrors,
+  // more middleware
+  (req, res) => res.status(200).json(res.locals.errors)
+);
+
+module.exports = router;
```
Comment (dlande000) [inline server/routers/errorRouter.js]: See https://github.com/oslabs-beta/Kubiquity/pull/12/files#r664688109

File: server/controllers/errorController.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+const HEADERS = [
+  'NAMESPACE',
+  'LAST SEEN',
+  'TYPE',
+  'REASON',
+  'OBJECT',
+  'MESSAGE'
+];
+
+const errorController = {};
```
Comment (dlande000) [inline server/controllers/errorController.js]: Change to `errorsController`: https://stackoverflow.com/questions/646951/singular-or-plural-controller-and-helper-names-in-rails

File: server/routers/errorRouter.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+const express = require('express');
+const errorController = require('../controllers/errorController');
```
Comment (dlande000) [inline server/routers/errorRouter.js]: See https://github.com/oslabs-beta/Kubiquity/pull/12/files#r664691786 (and relevant for the rest of this file)

File: server/routers/errorRouter.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+const router = express.Router();
+
+//Get Request
```
Comment (dlande000) [inline server/routers/errorRouter.js]: Delete comment.

File: server/routers/errorRouter.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+router.get(
+  '/',
+  // errorController.clearErrors,
+  errorController.queryErrors,
+  errorController.formatErrors,
+  errorController.saveErrors,
+  errorController.getErrors,
+  // more middleware
```
Comment (dlande000) [inline server/routers/errorRouter.js]: Delete comment.

File: server/models/errorsModel.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
+const mongoose = require('mongoose');
```
Comment (dlande000) [inline server/models/errorsModel.js]: change to: `const { Schema, connect, model } = require('mongoose')`, convert elsewhere in file as needed, and double check it works.

File: server/controllers/errorController.js [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+      log: `errorController.getErrors: ERROR: ${err}`,
+      message: {
+        err: 'Error occurred while retrieving errors. Check server logs for more information.',
+      }
+    })
+  }
+}
+errorController.clearErrors = async (req, res, next) => {
+  try {
+    await Error.deleteMany();
```
Comment (dlande000) [inline server/controllers/errorController.js]: See https://github.com/oslabs-beta/Kubiquity/pull/12/files#r664684205

File: server/controllers/errorController.js [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    for (let i = 0; i < formattedError.length; i++) {
+      const obj = {
+        namespace: formattedError[i][0],
+        lastSeen: formattedError[i][1],
+        type: formattedError[i][2],
+        reason: formattedError[i][3],
+        object: formattedError[i][4],
+        message: formattedError[i][5],
+      }
+      Error.create(obj)
```
Comment (dlande000) [inline server/controllers/errorController.js]: See https://github.com/oslabs-beta/Kubiquity/pull/12/files#r664684205

File: server/controllers/errorController.js [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+  'OBJECT',
+  'MESSAGE'
+];
+
+const errorController = {};
+
+//GOAL: Get output from kubectl command, possibly send to another middleware for formatting and processing data, store in mongodb
+// --field-selector type=Failure
+//Get output from kubectl command and store in a array with each row being contained in each index
+errorController.queryErrors = async (req, res, next) => {
```
Comment (dlande000) [inline server/controllers/errorController.js]: See: https://github.com/oslabs-beta/Kubiquity/pull/12/files#r664691786

File: server/controllers/errorController.js [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+      if (j === headersIndexArray.length - 2) {
+        subArray.push(array[i].slice(headersIndexArray[j + 1], array[i].length).trim())
+      }
+    }
+    arr.push(subArray);
+  }
+  return arr;
+}
+
+errorController.formatErrors = async (req, res, next) => {
```
Comment (dlande000) [inline server/controllers/errorController.js]: See: https://github.com/oslabs-beta/Kubiquity/pull/12/files#r664691786

File: server/controllers/errorController.js [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+  namespace,
+  lastSeen,
+  type,
+  reason,
+  object,
+}
+
+*/
+
+errorController.saveErrors = async (req, res, next) => {
```
Comment (dlande000) [inline server/controllers/errorController.js]: See: https://github.com/oslabs-beta/Kubiquity/pull/12/files#r664691786

File: server/controllers/errorController.js [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+    return next({
+      log: `errorController.saveErrors: ERROR: ${err}`,
+      message: {
+        err: 'Error occurred while saving errors. Check server logs for more information.',
+      }
+    })
+  }
+};
+
+errorController.getErrors = async (req, res, next) => {
```
Comment (dlande000) [inline server/controllers/errorController.js]: See: https://github.com/oslabs-beta/Kubiquity/pull/12/files#r664691786

File: server/controllers/errorController.js [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+  try {
+    res.locals.errors = await Error.find({});
```
Comment (dlande000) [inline server/controllers/errorController.js]: See: https://github.com/oslabs-beta/Kubiquity/pull/12/files#r664684205

File: server/controllers/errorController.js [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+//Store an array of the index of where each header in HEADERS array begins as well as the array length
+function headersIndex(array) {
+  let arr = [];
+  for (let i = 0; i < HEADERS.length; i++) {
+    arr.push(array.indexOf(HEADERS[i]));
+  }
+  //arr.push(array.length);
+  console.log('headersIndex', arr);
+  return arr;
+}
```
Comment (dlande000) [inline server/controllers/errorController.js]: If I'm reading this function correctly, we are creating an array of indices, right? And those values are hard-coded at the top. It seems like a better way to do this is to remove this function, and in `errorArrayConverter`, just run the second for loop while j < 5 (6 - 1)

File: server/controllers/errorController.js [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+
+//Converts ErrorList array of strings to nested array
+function errorArrayConverter(array) {
```
Comment (dlande000) [inline server/controllers/errorController.js]: Move this function declaration to line 12 and convert to arrow function (assuming that doesn't provoke an error).

File: server/controllers/errorController.js [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+  let arr = [];
```
Comment (dlande000) [inline server/controllers/errorController.js]: Use const.

File: server/controllers/errorController.js [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+  let headersIndexArray = headersIndex(array[0])
+  for (let i = 0; i < array.length; i++) {
+    let subArray = [];
```
Comment (dlande000) [inline server/controllers/errorController.js]: Use const.

File: server/controllers/errorController.js [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+    for (let j = 0; j < headersIndexArray.length - 1; j++) {
+      subArray.push(array[i].slice(headersIndexArray[j], headersIndexArray[j + 1]).trim())
```
Comment (dlande000) [inline server/controllers/errorController.js]: Could you please break `array[i].slice(headersIndexArray[j], headersIndexArray[j + 1]` into variables? I don't know what I'm trying to read here.

File: server/controllers/errorController.js (hunk 20 above)
Comment (dlande000) [inline server/controllers/errorController.js]: Pick clearer variable name; what type of data is stored in this array?

File: server/controllers/errorController.js (hunk 21 above)
Comment (dlande000) [inline server/controllers/errorController.js]: Pick clearer variable name—what type of data is being stored in this subArray?

File: server/controllers/errorController.js [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+      if (j === headersIndexArray.length - 2) {
+        subArray.push(array[i].slice(headersIndexArray[j + 1], array[i].length).trim())
```
Comment (dlande000) [inline server/controllers/errorController.js]: Same issue as https://github.com/oslabs-beta/Kubiquity/pull/12/files#r664702863

File: server/controllers/errorController.js [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+errorController.saveErrors = async (req, res, next) => {
+  try {
+    let formattedError = res.locals.formattedError;
+    // 1. Iterate through all our errors
+    // 2. save each instance of error; this will by async work
+    // 3. once all errors are saved, we want to find and return up all errors
+    // 4. save those errors to res.locals
+    // 5. hit that next piece of sweet sweet middleware
+    for (let i = 0; i < formattedError.length; i++) {
+      const obj = {
```
Comment (dlande000) [inline server/controllers/errorController.js]: Before `obj`, consider adding the following code: 

`const [
  namespace,
  lastSeen,
  type,
  reason,
  object,
  message
] = formattedError[i]`

Comment (dlande000) [reply server/controllers/errorController.js]: If the above is added, change your object to:

`{ namespace, lastSeen, type, reason. object, message }`

Comment (dlande000) [inline server/controllers/errorController.js]: Pick descriptive variable name instead of `obj`.

File: server/controllers/errorController.js [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+        object: formattedError[i][4],
+        message: formattedError[i][5],
+      }
+      Error.create(obj)
+    }
+    //res.locals.error
+
+    // const errors = Errors.findBy({});
+    // console.log(errors);
+    return next();
```
Comment (dlande000) [inline server/controllers/errorController.js]: How are we controlling for async work here? It seems like we would move to the next piece of middleware once we're done iterating through formattedError (is this one error or many??) This seems like potentially a good place to use `Promise.all`— @JefZheng @Hydroelectric29 any thoughts on that?

File: server/controllers/errorController.js [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+    }
+    arr.push(subArray);
+  }
+  return arr;
+}
+
+errorController.formatErrors = async (req, res, next) => {
+  try {
+    let headersIndexArray = headersIndex(res.locals.errorList[0])
+    let formattedError = errorArrayConverter(res.locals.errorList);
```
Comment (dlande000) [inline server/controllers/errorController.js]: use const.

Comment (dlande000) [reply server/controllers/errorController.js]: Should this be formattedErrors??

**Comment by dlande000** [comment]
Big takeaways: 

1. Our controller should be plural: all `errorController` should be `errorsController`
2. Our routes should be plural: all `errorRoutes` should be `errorsRoutes`
3. I'm having a hard time understanding the logic of `errorArrayConverter`—could you please break some work into descriptive variables so that I have a better sense of what's going on without a lot of reading and parsing? 
4. Mainly in `errorController`, but also the other files, please delete any comments or console logs that are no longer in use. Any comments that are meant to guide future development or to explain confusing bits of code are good, but all other comments should go.

**Comment by dlande000** [comment]
Screen Shot 2021-07-06 at 12 33 13 PM
once changes are made, please click the rereview icon for my name and I'll look over again.

File: server/controllers/errorController.js (hunk 18 above)
Comment (dlande000) [reply server/controllers/errorController.js]: After spinning up K8s and seeing the console log, I better understand the need of this helper function, so I believe my above comment can be disregarded.

**Review by Hydroelectric29 — APPROVED** [review]

**Review by dlande000 — APPROVED** [review]

**Pull request merged**